provider/filepath.go

```diff
@@ -0,0 +1,99 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package provider
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+type Filepath interface {
+	Bucket() string
+	Prefix() string
+	Path() string
+	FullPathWithBucket() string
+	FullPathWithoutBucket() string
+	ParseFullPath(path string) error
+}
+
+func NewFilepath(storeType string, bucket string, prefix string, path string) (Filepath, error) {
+	switch storeType {
+	case "s3":
+		return &S3Filepath{
+			genericFilepath: genericFilepath{
+				bucket: bucket,
+				prefix: prefix,
+				path:   path,
+			},
+		}, nil
+	default:
+		return nil, fmt.Errorf("unknown store type '%s'", storeType)
+	}
+}
+
+type genericFilepath struct {
```

Comment:
Make sure this handles trailing / vs no trailing / well since I know we've had a handful of bugs come from stuff like that

Reply:
Could also rename this to filePath. Also the idea of a bucket isn't necessarily generic for all file paths right?